locksmith/src/controllers/hookController.ts

```diff
@@ -15,178 +26,141 @@ const Hub = z.object({
-
-      const createdHook = await Hook.create({
-        expiration,
-        topic: hub.topic,
-        callback: hub.callback,
-        mode: hub.mode,
-        secret: hub.secret,
-        network: params.network,
-        lock: params.lock,
-      })
-
-      return createdHook
-    }
-
-    if (hook.mode !== hub.mode) {
-      hook.mode = hub.mode
+  async handle(request: SubscribeRequest, response: Response) {
+    const network = this.getNetwork(request.params.network)
+    if (!network) {
+      return response.status(404).send('Unsupported Network')
     }
-
-    if (hub.lease_seconds) {
-      hook.expiration = getExpiration(hub.lease_seconds)
+    const result = await Hub.safeParseAsync(request.body.hub)
+    if (!result.success) {
+      return response.status(400).send({
+        hub: {
+          mode: request.body.hub.mode,
+          topic: request.body.hub.topic,
+          reason: result.error.message,
+        },
+      })
+    } else {
+      // Send the accepted request to the subscriber and then validate the intent of the subscriber as well as persist the subscription.
+      response.status(202).send('Accepted')
+      try {
+        await this.verifySubscriber(request)
+        await this.updateHook(request)
+        return
+      } catch (error) {
+        logger.error(error.message)
+        // eslint-disable-next-line
+        return
+      }
     }
+  }
 
-    if (hub.secret) {
-      hook.secret = hub.secret
+  // Create an expiration date based on lease_seconds
+  getExpiration(leaseSeconds?: number) {
+    const limit = leaseSeconds ?? this.options.leaseSeconds.default
+    if (limit > this.options.leaseSeconds.limit) {
+      throw new Error("Lease seconds can't be greater than 90 days")
     }
+    return new Date(Date.now() + limit * 1000)
+  }
 
-    await validHubIntent(hub)
+  // Get the correct network config based on id
+  getNetwork(id: string) {
+    return networks[id]
+  }
 
-    await hook.save()
-    return hook
-  } catch (error) {
-    logger.error(error.message)
-    return error
+  // Generate a random string of a buffer of size provided.
+  getChallege(size = 20) {
+    return crypto.randomBytes(size).toString('hex')
   }
-}
 
-// Check if the subscriber request is valid or not. This will post a challenge to subscriber to confirm the intent of the request.
-export async function validHubIntent(hub: z.infer<typeof Hub>) {
-  const challenge = crypto.randomBytes(20).toString('hex')
-  const result = await fetch(hub.callback, {
-    method: 'POST',
-    headers: {
-      'Content-Type': 'application/json',
-    },
-    body: JSON.stringify({
+  // Verify the subscriber intent by having them echo back the challenge, hub topic, and hub callback. If they match, we can subscribe.
+  async verifySubscriber(request: SubscribeRequest) {
+    const { hub } = request.body
+    const challenge = this.getChallege()
+
+    const body = {
       hub: {
         ...hub,
         challenge,
```

Comment:
We need to add the expiration lease here

https://www.w3.org/TR/websub/#x5-3-hub-verifies-intent-of-the-subscriber